### PR TITLE
feat: Add setValueAtPath template function

### PR DESCRIPTION
**Pull request by twpayne**
Replaces #2530.

### CODE REVIEW

**Review by halostatue — APPROVED** [review]

File: assets/chezmoi.io/docs/reference/templates/functions/setValueAtPath.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# `setValueAtPath` *path* *value* *dict*
+
+`setValueAtPath` modifies *dict* to set the value at *path* to *value* and
+returns *dict*. *path* can be either a string containing a `.`-separated list of
+keys or a list of keys. It is an alternative to [sprig's `set`
+function](http://masterminds.github.io/sprig/dicts.html) with a different
+argument order that supports pipelining.
```
Comment (halostatue) [inline assets/chezmoi.io/docs/reference/templates/functions/setValueAtPath.md:7]: We should probably document that `setValueAtPath "root.key.bar.baz" "quux"` will *absolutely* error out given the following input:

```json
{
  "root": {
    "key": {}
  }
}
```

It *could* be nice to have a `createOrSetValueAtPath` variant which would create `"bar": { "bar": "quux" }`.

File: assets/chezmoi.io/docs/reference/templates/functions/setValueAtPath.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+!!! example
+
+    ```
+    {{ dict | setValueAtPath "key1" "value1" | setValueAtPath "key2.nestedKey" "value2" | toJson }}
+    {{ dict | setValueAtPath (list "key2" "nestedKey") "value2" | toJson }}
+    ```
```
Comment (halostatue) [inline assets/chezmoi.io/docs/reference/templates/functions/setValueAtPath.md:14]: Should we show the sample output?

More to the point, would `cat '{}' | chezmoi execute-template "{{ .chezmoi.stdin | fromJson | setValueAtPath "key1" "value" }}` work, as that would make it easier to demonstrate.

**Comment by felipecrs** [comment]
This is awesome!

File: assets/chezmoi.io/docs/reference/templates/functions/setValueAtPath.md (hunk 1 above)
Comment (twpayne) [reply assets/chezmoi.io/docs/reference/templates/functions/setValueAtPath.md:7]: This doesn't error, it creates the structure as needed:

```console
$ ./chezmoi --version
chezmoi version dev, commit 27c4ed8df22885d9ea49b4c9d0829d9b2097aebe, built at 2022-11-01T20:18:58Z
$ ./chezmoi execute-template '{{ dict "root" (dict "key" dict) | toPrettyJson }}'
{
  "root": {
    "key": {}
  }
}
$ ./chezmoi execute-template '{{ dict "root" (dict "key" dict) | setValueAtPath "root.key.bar.baz" "quux" | toPrettyJson }}'
{
  "root": {
    "key": {
      "bar": {
        "baz": "quux"
      }
    }
  }
}                                                                                                                                                                                                                                                                                                                                         
```

File: assets/chezmoi.io/docs/reference/templates/functions/setValueAtPath.md (hunk 2 above)
Comment (twpayne) [reply assets/chezmoi.io/docs/reference/templates/functions/setValueAtPath.md:14]: All the other template function examples don't include the output.

File: assets/chezmoi.io/docs/reference/templates/functions/setValueAtPath.md (hunk 1 above)
Comment (twpayne) [reply assets/chezmoi.io/docs/reference/templates/functions/setValueAtPath.md:10]: I've updated the documentation to document the create behavior. I don't think we need to rename the function as the name is already quite unwieldy.

Comment (halostatue) [reply assets/chezmoi.io/docs/reference/templates/functions/setValueAtPath.md:10]: We only needed the function renamed if the behaviour was halt-y.

**Review by halostatue — APPROVED** [review]

**Pull request merged**